### PR TITLE
TypeScript: Convert `output`package

### LINKED ISSUE #12089: TypeScript: Convert `output` package

### PR DESCRIPTION
**Pull request by amovar18**
## Context

This PR aims to convert the `@googleforcreators/output` package from JavaScript to TypeScript.

## Summary

This PR creates required types for `@googleforcreators/output` package and converts the JavaScript files to TypeScript file.

## To-do

- [ ] Remove various comments regarding `eslint` disable and enable which have been placed since `@googleforcreators/elements`, `@googleforcreators/masks`, `@googleforcreators/element-library` and `@googleforcreators/animation` haven't yet been converted to TypeScript.
- [ ] Convert Test cases to TypeScript.
- [ ] Need to figure out `amp-*` tags types.

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12089

### CODE REVIEW

File: packages/output/package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
     }
   },
   "main": "dist/index.js",
+  "types": "dist-types/index.d.ts",
   "module": "dist-module/index.js",
   "source": "src/index.js",
```
Comment (swissspidy) [inline packages/output/package.json]:
```suggestion
  "source": "src/index.ts",
```

File: packages/output/package.json [hunk 2]
```diff
@@ -48,6 +49,9 @@
     "@googleforcreators/rich-text": "*",
     "@googleforcreators/units": "*",
     "@googleforcreators/url": "*",
+    "@googleforcreators/types": "*",
+    "@types/react": "^18.0.21",
+    "@types/react-dom": "^18.0.6",
```
Comment (swissspidy) [inline packages/output/package.json]: We currently use React 17

```suggestion
    "@types/react": "^17.0.44",
    "@types/react-dom": "^17.017",
```

File: packages/types/src/element/textElement.ts [hunk 3]
```diff
@@ -37,6 +37,7 @@ export interface Font {
   styles?: FontStyle[];
   variants?: FontVariant[];
   fallbacks?: string[];
+  url?: string;
```
Comment (swissspidy) [inline packages/types/src/element/textElement.ts]: Since only self-hosted fonts have a `url` property, we might want to split this up into multiple types to clarify (system fonts, custom fonts, Google Fonts).

File: packages/types/src/element/productElement.ts [hunk 4]
```diff
@@ -18,7 +18,25 @@
...
+  productImages: ProductImage[];
+  productPrice: number;
+  productPriceCurrency: string;
+  productTitle: string;
+  productUrl: string;
+}
+export interface ProductElement extends Element {
+  product: Product;
+}
+export interface ShoppingAttachmentType {
```
Comment (swissspidy) [inline packages/types/src/element/productElement.ts]: We don't usually suffix types with "Type".

```suggestion
export interface ShoppingAttachment {
```

File: packages/output/src/element.tsx [hunk 5]
```diff
@@ -33,13 +31,13 @@ import {
...
 } from '@googleforcreators/masks';
-
 /**
  * Internal dependencies
  */
 import WithLink from './components/withLink';
+import type { OutputElementTyping } from './types';
 
-function OutputElement({ element, flags }) {
+function OutputElement({ element, flags }: OutputElementTyping) {
```
Comment (swissspidy) [inline packages/output/src/element.tsx]: We don't usually suffix types with "Typing".

Also, this type is only used by this specific component, so it can just live in this file here, it doesn't have to imported from `types.ts`

File: packages/output/src/components/withLink/index.tsx [hunk 6]
```diff
@@ -17,12 +17,15 @@
...
 import { withProtocol } from '@googleforcreators/url';
-import { StoryPropTypes } from '@googleforcreators/elements';
+/**
+ * Internal dependencies
+ */
+import type { WithLinkTyping } from '../../types';
 
-function WithLink({ element, children, ...rest }) {
-  const link = element.link || {};
+function WithLink({ element, children, ...rest }: WithLinkTyping) {
```
Comment (swissspidy) [inline packages/output/src/components/withLink/index.tsx]: Ditto here about "Typing" and importing.

I'd rename to `WithLinkProps` and just define it here

File: packages/fonts/package.json [hunk 7]
```diff
@@ -47,7 +47,7 @@
   },
   "main": "dist/index.js",
   "module": "dist-module/index.js",
-  "tyoes": "dist-types/index.d.ts",
```
Comment (swissspidy) [inline packages/fonts/package.json:50]: Good catch!

File: packages/types/src/element/element.ts [hunk 8]
```diff
@@ -65,4 +75,9 @@ export interface Element extends ElementBox {
   groupId?: string;
   border?: Border;
   borderRadius?: BorderRadius;
+  pageAttachment?: PageAttachment;
```
Comment (barklund) [inline packages/types/src/element/element.ts]: Elements can't have page attachments, only pages can.

File: packages/types/src/element/element.ts [hunk 9]
```diff
@@ -65,4 +75,9 @@ export interface Element extends ElementBox {
...
+  resource?: Resource;
+  overlay?: Pattern;
```
Comment (barklund) [inline packages/types/src/element/element.ts]: These are only properties of media elements, not all elements.

File: packages/types/src/element/element.ts [hunk 10]
```diff
@@ -65,4 +75,9 @@ export interface Element extends ElementBox {
...
+  backgroundColor?: Pattern;
```
Comment (barklund) [inline packages/types/src/element/element.ts]: This is only a property of shape and text elements, not all elements.

File: packages/types/src/element/element.ts [hunk 11]
```diff
@@ -65,4 +75,9 @@ export interface Element extends ElementBox {
...
+  backgroundTextMode?: string;
```
Comment (barklund) [inline packages/types/src/element/element.ts]: This is only a property of text elements, not all elements.

File: packages/types/src/story/page.ts [hunk 12]
```diff
@@ -35,4 +46,9 @@ export interface Page {
   backgroundColor: Pattern;
   type: 'page';
   groups: Groups;
+  id: string;
+  pageAttachment?: PageAttachment;
+  advancement: Advancement;
+  backgroundAudio: BackgroundAudio;
+  shoppingAttachment: ShoppingAttachmentType;
```
Comment (barklund) [inline packages/types/src/story/page.ts]: All of these are optional, no?

File: packages/output/src/types.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  backgroundAudio: BackgroundAudio;
+  id: string;
+}
+export interface PreloadResource {
+  url: string;
+  type: string;
+}
+export interface OutputElementTyping {
+  element: Element;
+  flags: Record<string, boolean>;
```
Comment (barklund) [inline packages/output/src/types.ts]: This (and below) could also just be `flags: Flags`, no?

**Review by barklund — COMMENTED** [review]
A few comments on the types.

File: packages/types/src/element/element.ts (hunk 8 above)
Comment (amovar18) [reply packages/types/src/element/element.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/types/src/element/element.ts (hunk 10 above)
Comment (amovar18) [reply packages/types/src/element/element.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/types/src/story/page.ts (hunk 12 above)
Comment (amovar18) [reply packages/types/src/story/page.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/types/src/element/element.ts (hunk 11 above)
Comment (amovar18) [reply packages/types/src/element/element.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/output/src/types.ts (hunk 13 above)
Comment (amovar18) [reply packages/output/src/types.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/types/src/element/element.ts (hunk 9 above)
Comment (amovar18) [reply packages/types/src/element/element.ts]: Fixed in https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/a1cea5e2504a0731e448b3e5895b18eb6562016d

File: packages/fonts/package.json [hunk 14]
```diff
@@ -47,7 +47,7 @@
...
+  "types": "dist-types/index.d.ts",
```
Comment (spacedmonkey) [inline packages/fonts/package.json:50]: Unrelated change?

File: packages/fonts/src/utils/getFontCSS.ts [hunk 15]
```diff
@@ -21,7 +21,11 @@
  * @return Font format string or null if the font is unsupported.
  */
 function getFontFormat(src: string): string | null {
-  const fileExtension = src.split(/[#?]/)?.[0]?.split('.')?.pop()?.trim();
+  const fileExtension: string | undefined = src
```
Comment (spacedmonkey) [inline packages/fonts/src/utils/getFontCSS.ts]: All these changes to the font package required? If not, lets add them to another PR.

File: packages/fonts/src/utils/getGoogleFontURL.ts [hunk 16]
```diff
@@ -30,7 +30,10 @@ import type { Font } from '../types';
  * @param [display] Valid font-display value, e.g. 'swap' or 'auto'. Default 'swap'.
  * @return Google Fonts embed URL.
  */
-function getGoogleFontURL(fonts: Font[], display: 'swap' | 'auto' = 'swap') {
+function getGoogleFontURL(
+  fonts: Font[],
+  display: 'swap' | 'auto' = 'swap'
+): string {
```
Comment (spacedmonkey) [inline packages/fonts/src/utils/getGoogleFontURL.ts]: Can't we just infer the return type here?

File: packages/output/src/element.tsx [hunk 17]
```diff
@@ -33,22 +31,20 @@ import {
...
-    backgroundColor,
-    backgroundTextMode,
-    overlay,
-  } = element;
+function OutputElement({ element, flags }: OutputElementTyping) {
+  let overlay;
+  if ('overlay' in element) {
+    overlay = element.overlay;
+  }
+  const { id, opacity, type, border, backgroundColor, backgroundTextMode } =
```
Comment (spacedmonkey) [inline packages/output/src/element.tsx]: Lint error, seem valid. here.

File: packages/output/src/utils/getLongestMediaElement.tsx [hunk 18]
```diff
@@ -17,22 +17,21 @@
...
  * Among all elements, returns the media element with the longest duration.
- *
- * @param {Array<Object>} elements List of elements.
- * @param {number} minDuration Duration that the minimum element must exceed in seconds.
- * @return {Object|undefined} Found element, or undefined if there are no media elements.
  */
-function getLongestMediaElement(elements, minDuration = 0) {
-  return elements
+function getLongestMediaElement(elements: Element[], minDuration = 0): Element {
+  const videoElements = elements as VideoElement[];
```
Comment (spacedmonkey) [inline packages/output/src/utils/getLongestMediaElement.tsx]:
```suggestion
function getLongestMediaElement(elements: VideoElement[], minDuration = 0): Element {
  return elements
```

**Comment by spacedmonkey** [comment]
@amovar18 This PR is a good start. There are a number of lint and type errors that need to be fixed. Until these are fixed, this PR will not be considered finished. There are also merge conflicts currently. Those should be resolved first. 

Having a quick look there seems to type errors missing for the following packages. 

@googleforcreators/url
@googleforcreators/patterns
@googleforcreators/animation
@googleforcreators/elements
@googleforcreators/masks
@googleforcreators/design-system
@googleforcreators/rich-text

File: packages/fonts/package.json (hunk 14 above)
Comment (swissspidy) [reply packages/fonts/package.json:50]: But a correct one. Fixes a typo.

**Comment by swissspidy** [comment]
That's OK. It's a noted to-do above. It was me who recommended opening a draft PR at this early stage.

**Comment by swissspidy** [comment]
@amovar18 Great work so far, thanks for your help!

If you don't mind, I will push some changes/fixes myself so we can get the ball rolling again on this one.

File: packages/fonts/src/utils/getFontCSS.ts (hunk 15 above)
Comment (amovar18) [reply packages/fonts/src/utils/getFontCSS.ts]: I dont think its required. Without this linter shows a lot of lints.

**Comment by amovar18** [comment]
@swissspidy yes sure you can push any fixes/improvements that you want to.

File: packages/types/src/element/productElement.ts (hunk 4 above)
Comment (amovar18) [reply packages/types/src/element/productElement.ts]: Fixed in [0f952b7](https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/0f952b7b62092961c0d19a90db729dec46711097)

File: packages/output/src/element.tsx (hunk 5 above)
Comment (amovar18) [reply packages/output/src/element.tsx]: Fixed in [0f952b7](https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/0f952b7b62092961c0d19a90db729dec46711097)

File: packages/output/src/components/withLink/index.tsx (hunk 6 above)
Comment (amovar18) [reply packages/output/src/components/withLink/index.tsx]: Fixed in [0f952b7](https://github.com/GoogleForCreators/web-stories-wp/pull/12382/commits/0f952b7b62092961c0d19a90db729dec46711097)

**Comment by amovar18** [comment]
@swissspidy What to do regarding the `amp-*` tags? Any ideas?

**Comment by swissspidy** [comment]
@amovar18 I am currently working on some fixes for this PR, please hold off for a bit.
